### PR TITLE
ci: include more files in checksum generation

### PR DESCRIPTION
**Pull request by jhchabran**
Follow-up to https://github.com/sourcegraph/sourcegraph/pull/38940 

We had a quick sync with @valerybugakov and we agreed that it would be safer to: 

- cover the missing files that weren't used to build up the index
- add a banner making it explicit that the bundle has been cached for the build

CleanShot 2022-07-19 at 14 37 02@2x

On his side, @valerybugakov will add in https://github.com/sourcegraph/sourcegraph/issues/39034 a way to check if the bundle came from the cache or not, so we could quickly see if that's the case or not. 

💡 Keep in mind that the cache we're seeing here, **only applies to the server image**, all others are not using the cache. This limits the blast radius of a possible issue with this optimization.

## Test plan

Tested in CI again (cache miss + hit)

### CODE REVIEW

File: enterprise/cmd/server/pre-build.sh [hunk 1]
```diff
@@ -3,6 +3,27 @@
...
+  tmpfile=$(mktemp)
+  # shellcheck disable=SC2064
+  trap "rm \"$tmpfile\"" EXIT
+
+  {
+    find "./client" "./ui" "yarn.lock" -type f -exec sha1sum {} \;
+    find . -maxdepth 1 -type f -name "*.js" -exec sha1sum {} \;
+    find . -maxdepth 1 -type f -name "*.ts" -exec sha1sum {} \;
+    find . -maxdepth 1 -type f -name "*.json" -exec sha1sum {} \;
+  }>> "$tmpfile"
```
Comment (jhchabran) [inline enterprise/cmd/server/pre-build.sh:16]: @valerybugakov, @eseliger this cover much more now.

**Comment by jhchabran** [comment]
> Looking at it shows that it's not up to date: babel.config.js and svgo.config.js are missing.

Fixed

> There're several non-root files that we rely on to generate type definitions. Most probably, they do not affect the bundle, but we were discussing generating Apollo Client hooks based on the GraphQL schema someday (...): 

> The dangerous part of the setup is the implicit dependency on a hardcoded list of files. The codebase evolves daily, and we might introduce a new dependency easily. Currently, we don't have a way to get all inputs for the client build programmatically to automate this. I can see a scenario where we would use the outdated bundle because an important file that didn't make it to the list of cache-key dependencies was changed. As far as I understand, this inconsistency will be hard to spot and debug.

I'm repeating some of the points we covered with @valerybugakov this morning in a quick call. 

Yes, this is dangerous, but a/ we're limiting the blast radius by keeping this optimization only for the server image, and /b it is the biggest bottleneck in the CI. It's just making that step faster, but also speeds up the whole build by around 6/7 minutes. Which is absolutely huge. 

Therefore, given the outcome, I think we shouldn't be scared of cutting some corners we can take the risk of having to remember to update this if we change something in the build system. 

@valerybugakov brought up during our sync a great idea, which is to make it, first, easy to see if your build was made with the cache, and second make this easily verifiable in the client console, so we could see if the cache wasn't invalidated where it should have been, by being able to ask the client code which commit it was build with. This is to be done by @valerybugakov in TODO. 

> Achieving similar benefits safely requires an explicit binding between tasks and cache-key dependencies. An introduction of the tasks manager, who would track inputs of all dependent tasks, tasks graph, and calculate cache keys based on them, is one of the standard solutions. (e.g., turborepo, nx, bazel)

I totally agree. My main problem is that we have been postponing such efforts for more than 6 months now. While I personally plan to work on Bazel, I don't think we should forbid ourselves to apply a bandage just because a big improvement is somewhere on the roadmap.

> Let me know if I'm overly cautious here or miss any context 🙂
> @jasongornall

I very much appreciate your carefulness here, the consequences of a cache incorrectly invalidated could be very annoying. Which is why I think it's important that both our team are onboard with this and aware on how it works. 

What I propose is that we keep observing this for a few weeks and see if we stumble on some issues. Again, I repeat, this optimization is not affecting cloud, nor dotcom or anything else other than the single docker image deployment.

**Review by eseliger — APPROVED** [review]
should we disable caching on a release branch? I think that'd make me feel _much_ better and it's a net gain to me :) regardless, approve.

**Comment by jhchabran** [comment]
> should we disable caching on a release branch? 

This is a great idea, I'll pull another PR 🙏

**Review by valerybugakov — APPROVED** [review]
Awesome writeup, @jhchabran! Thanks for documenting all the context here based on our discussion. Agree with the assessment, and the shortcut feels much safer with this beautiful, in-your-face banner 🙌

File: enterprise/cmd/server/pre-build.sh [hunk 2]
```diff
@@ -20,23 +41,37 @@ else
...
-    echo -e "ClientBundle 🔥 Cache hit: $cache_key"
+    echo -e "~~~ ClientBundle 🔥 Cache hit: $cache_key"
+    # Getting the cached bundle
     aws s3 cp --profile buildkite --endpoint-url 'https://storage.googleapis.com' --region "us-central1" "s3://sourcegraph_buildkite_cache/$cache_key" "./"
     bsdtar xzf "$cache_file"
     rm "$cache_file"
+
+    # Retrieving the cache description
+    aws s3 cp --profile buildkite --endpoint-url 'https://storage.googleapis.com' --region "us-central1" "s3://sourcegraph_buildkite_cache/$cache_desc_key" "./"
+    echo -e "ClientBundle 🔥 Cache hit: \`$cache_key\`\n\n$(cat "$cache_desc_file")" | ./enterprise/dev/ci/scripts/annotate.sh -m -t info
```
Comment (bobheadxi) [inline enterprise/cmd/server/pre-build.sh:60]: Should we just use the pipeline generator annotation API instead?

Comment (bobheadxi) [reply enterprise/cmd/server/pre-build.sh:60]: I think @burmudar was planning on working on removing the `annotate.sh` script entirely

Comment (jhchabran) [reply enterprise/cmd/server/pre-build.sh:60]: Yes, but I wanted to have info, not warnings. Can we do it with the annotation API ?

Comment (bobheadxi) [reply enterprise/cmd/server/pre-build.sh:60]: yes, testing it out here: https://github.com/sourcegraph/sourcegraph/pull/39080

Comment (bobheadxi) [reply enterprise/cmd/server/pre-build.sh:60]: one disadvantage of the API right now is that annotation levels are command-wide, but I think that's okay for now

**Pull request merged**